custom_metrics/markup.js

```diff
@@ -0,0 +1,580 @@
+        // https://developers.google.com/search/reference/robots_meta_tag
+        var validNodes = document.querySelectorAll('span[data-nosnippet], div[data-nosnippet], section[data-nosnippet]');
+        var allNodes = document.querySelectorAll('[data-nosnippet]');
+        return { valid: validNodes.length, wrong_tag_type: allNodes.length - validNodes.length};
+      }
+      catch(e) {
+        return logError("data_nosnippet", e);
+      }
+    })(),
+
+    // markup info 
+    // Used by: Markup
+    'obsolete_elements': (() => {  
+      try {   
+        // https://html.spec.whatwg.org/multipage/obsolete.html#non-conforming-features
+        // Array.from(document.querySelectorAll('dfn[data-dfn-type] code')).map(e => e.innerText).join(',')
+        let result = {};
+        let obsoleteNodes = [...document.querySelectorAll('applet,acronym,bgsound,dir,frame,frameset,noframes,isindex,keygen,listing,menuitem,nextid,noembed,plaintext,rb,rtc,strike,xmp,basefont,big,blink,center,font,marquee,multicol,nobr,spacer,tt')];
+
+        obsoleteNodes.forEach((n) => {
+          let t = n.tagName.toLowerCase();
+            if (result[t])
+              result[t]++;
+            else
+              result[t] = 1;
+        });
+
+        return result;
+      }
+      catch(e) {
+        return logError("obsolete_elements", e);
+      }
+    })(),
+
+    // svg use
+    // Used by Markup
+    'svgs': (() => {   
+      try {   
+        let result = {};
+
+        result.svg_element_total = document.querySelectorAll('svg').length;
+        result.svg_img_total = document.querySelectorAll('img[src*=".svg"]').length;
+        result.svg_object_total = document.querySelectorAll('object[data*=".svg"]').length;
+        result.svg_embed_total = document.querySelectorAll('embed[src*=".svg"]').length;
+        result.svg_iframe_total = document.querySelectorAll('iframe[src*=".svg"]').length;
+
+        result.svg_total = result.svg_element_total+result.svg_img_total+result.svg_object_total+result.svg_embed_total+result.svg_iframe_total;
+
+        return result;
+      }
+      catch(e) {
+        return logError("svgs", e);
+      }
+    })(),
+
+    // app
+    // Used by Markup
+    'app': (() => {   
+      try {   
+        let result = {};
+
+        //  `<div id="app">` 
+        result.app_id_present = !!document.getElementById("app");
+
+        // `<meta name="theme-color">` 
+        result.meta_theme_color = document.querySelector('meta[name="theme-color"]')?.getAttribute('content') ?? null;
+
+        return result;
+      }
+      catch(e) {
+        return logError("app", e);
+      }
+    })()
+  };
+}
+catch(e) { // probably an un caught exception
+  logError("general", "Failed to create the custom metrics object", e);
+  _custom_metrics = {};
+}
+
+if (!_custom_metrics) { // should not be possible
+  _custom_metrics = {};
+  logError("general", "custom metrics object was missing");
+}
```

Comment:
If it's not possible, why are we including it?

Reply:
> If it's not possible, why are we including it?

Because I'm careful. I've done a lot of coding in my time and seen code hit that I didn't think was possible. 

Like I do try catches inside the logError. JavaScript exception objects are horrible.

I can remove it. In this case the result would most likely be the custom metric is blank instead of the message.

Reply:
Removing it.